### ISSUE TITLE
What happens when the disagreement number set by the group is larger than the group itself? [Discussion]

### ISSUE THREAD

**Issue by mmbotelho** [okTurtles/group-income#893]
### Problem

As @taoeffect said on Slack:
> Eg: you set it (disagreement number) to 5, your group has shrunk to 4 members. Impossible to prevent proposals from passing. Everyone could say no and it will still pass.

This is a problem when:
- Groups increase in size;
- Groups shrink;
- The group creator sets a very high number on the group creation flow and the first proposal created in the group has a disagreement number higher than the group size;
- Someone creates a proposal to change the disagreement number to something higher than the group size.

### Solution

There are a number of possible solutions for this problem:
- Proposals fail if either the absolute threshold you set is lower than the group size OR 50% of the group has voted no;
- Prevent users from creating proposals when the disagreement number they set is higher than the total group size and prompting them to create a proposal to change voting system OR the disagreement number;
- Even if users set a disagreement number of 30, and their group has 20 people, we automatically adjust it to 19 (which is closer to their initial intent);
- Leave it as-is and let users assume responsibility for whatever number they set there;

These are the solutions I can think of at the top of my head, let me know if you can think of something else!

**Comment by sandrina-p** [comment]
My vote goes to:
- Option 5: Change disagreement rule to be percentage based, avoiding this tricky scenario or any other possible bugs/misunderstandings/edge-cases.

**Comment by taoeffect** [comment]
> revent users from creating proposals when the disagreement number they set is higher than the total group size and prompting them to create a proposal to change voting system OR the disagreement number;

This situation would result in a Catch-22, as the proposal won't be able to pass, but you need to make it to change the threshold.

> Even if users set a disagreement number of 30, and their group has 20 people, we automatically adjust it to 19 (which is closer to their initial intent);

The code is not designed in a way to automatically adjust the threshold like that. We would have to add special threshold-checking code anytime someone leaves, and that has "code smell".

> Leave it as-is and let users assume responsibility for whatever number they set there;

We can't leave users stuck in a hopeless situation, that's poor UX.

> Proposals fail if either the absolute threshold you set is lower than the group size OR 50% of the group has voted no;

This (slightly worded differently) is the solution I think makes the most amount of sense to me, and is based around [our first blog post](https://groupincome.org/2016/06/what-makes-a-good-voting-system/) on what makes a good voting system.

![](https://groupincome.org/wp-content/uploads/2016/06/voting-v6.2.png)

To me think makes intuitive sense, to make the rule be `threshold = MIN(50% of current group size, threshold)`.

In other words, if the threshold for `RULE_DISAGREEMENT` is set to 4, and the group size is 10, then `MIN(4, 5) => 4` disagreements will be enough to block a proposal.

This way, `RULE_DISAGREEMENT` never allows users to place themselves in a precarious situation (the one we currently display a visual warning for in `RULE_PERCENTAGE`).

If for some reason they do want to play with fire and use sub-majority threshold rules, then they can switch to `RULE_PERCENTAGE` which will allow them to do that.

**Comment by mmbotelho** [comment]
> This way, RULE_DISAGREEMENT never allows users to place themselves in a precarious situation (the one we currently display a visual warning for in RULE_PERCENTAGE).

> If for some reason they do want to play with fire and use sub-majority threshold rules, then they can switch to RULE_PERCENTAGE which will allow them to do that.

Isn't it weird that we allow users to "play with fire" with RULE_PERCENTAGE and not RULE_DISAGREEMENT?

Like you said on Slack:
> i think, @Margarida, it's like asking, can you use a kitchen knife to stab yourself? my answer is yes, that is possible

It's a weird experience to have liberty in one system and restriction in the other. We either limit their choices on both voting systems or give them all the options in both.

**Comment by taoeffect** [comment]
I don’t think it’s weird, because these are two fundamentally different voting rules. What is weird to me is thinking there should be some kind of equivalence between them. There is no reason that should be.

Example: another possible voting rule is whatever the group creator/admin says. Voting rules are fundamentally different, and in respecting those differences we are able to make them each work well, and offer meaningful choice to users.

**Comment by mmbotelho** [comment]
Ok so there's 2 different things here:

1. The voting system mechanics;
2. What we allow users to do with said voting system (how restrictive or liberal we are).

Number 1 - You are absolutely right that the voting systems are very different, as they should be, or there would be no reason to have multiple voting systems. 

Number 2 however, is a different thing. There are 2 positions we can adopt:
- We can be very protective and restrict users' freedom of choice, which sometimes means we override their settings with the ones we think work best;
- We can be very liberal and let users choose to do whatever they want even if we, designers and developers, don't think that's a good choice. We can warn them and guide them, but ultimately, they have the final say.

And what I'm saying is that yes, we can have different voting systems, but no, it doesn't make to have the 2 positions mentioned above at the same time, because that's confusing for users. 

I hope it's clearer now

**Comment by mmbotelho** [comment]
I'm ok with either approach there, I just think it is important that we are consistent.

**Comment by taoeffect** [comment]
I don't see inconsistencies in making each voting rule work the best way we can make it work.

There are unique "gotchas" to `RULE_DISAGREEMENT` that simply don't exist in `RULE_PERCENTAGE`, and so we need to take special care to address those gotchas.

`RULE_PERCENTAGE` will never leave you in an impossible situation. `RULE_DISAGREEMENT`, without this adjustment I described, will.

**Comment by mmbotelho** [comment]
Well, I think we have a tie here then. Maybe it would be helpful to hear what others think? 

@dotmacro @sandrina-p @pieer do you have any thoughts on this one? Me and @taoeffect could use your input here.

**Comment by taoeffect** [comment]
Yes, I think it would be helpful to hear what others think. 👍 

I also forgot to say that regarding this:

> - We can be very liberal and let users choose to do whatever they want even if we, designers and developers, don't think that's a good choice. We can warn them and guide them, but ultimately, they have the final say.

That is the approach I'm leaning toward. We remain liberal because users can always choose to switch to `RULE_PERCENTAGE` if they don't like how `RULE_DISAGREEMENT` works.

Also, we need a working solution. I described the problems I see with the other solutions mentioned. So I haven't heard of a solution that really works well for `RULE_DISAGREEMENT` other than `threshold = MIN(threshold, half-of-group)`.

**Comment by sandrina-p** [comment]
Hi guys. I've been kinda out of this discussion around `RULE_DISAGREEMENT` because I don't even see how it's different (what are the benefits) compared to `RULE_PERCENTAGE`. 😶

Let's imagine a group with 10 members with `RULE_PERCENTAGE` set to 70%.
If 3 people vote against, the proposal gets refused. AFAIU that's exactly the same as setting `RULE_DISAGREEMENT` to 3 but in a smarter way because its value is adjusted based on the group size, avoiding the problems mentioned in this thread.

But if I had to choose from one of the options I would go with: 

> Leave it as-is and let users assume responsibility for whatever number they set there;

>  We can warn them and guide them, but ultimately, they have the final say.

Let's trust the people who use GI. The less rules we set, the better. Less code smells, less complexity. The responsibility is on the users side.

**Comment by taoeffect** [comment]
> Leave it as-is and let users assume responsibility for whatever number they set there;

Can't go with that for the reasons stated [here](https://github.com/okTurtles/group-income-simple/issues/893#issuecomment-613184751).

> Let's trust the people who use GI.

Our users are trusting us to do a good job and not deliver to them code that breaks.

**Comment by taoeffect** [comment]
This is also an interesting case study where voting simply doesn't work. I have not heard of a better solution so far than `threshold = MIN(threshold, half-of-group)`, and it doesn't matter how many people tell me otherwise, I will not approve PRs containing half-baked solutions that damage user experience. If I did, it would mean I wasn't doing my job as project manager. :-\

**Comment by taoeffect** [comment]
Regarding:

> Let's imagine a group with 10 members with `RULE_PERCENTAGE` set to 70%.
If 3 people vote against, the proposal gets refused. AFAIU that's exactly the same as setting `RULE_DISAGREEMENT` to 3 but in a smarter way because its value is adjusted based on the group size, avoiding the problems mentioned in this thread.

1. It may look like the same thing, but these are fundamentally different voting rules — one based on percentage, and one based on absolute number of objections. Very different.
2. Likewise, it is not smarter, just different. The problem described in this thread is easily fixed by doing `threshold = MIN(threshold, half-of-group)` for `RULE_DISAGREEMENT`, and suddenly there is no problem anymore.

So, unless anyone has any better suggestions (meaning: they don't break the app), that's what the math will have to be.

**Comment by sandrina-p** [comment]
> I have not heard of a better solution so far than `threshold = MIN(threshold, half-of-group)`

We told you other solutions and each one has cons and pros. It's kinda impossible to know what's the best solution when we are creating a new system with zero user research.

>  I will not approve PRs containing half-baked solutions that damage user experience. 

So, your solution can't go neither because doing `MIN(threshold, half-of-group)` is confusing as @mmbotelho mentioned before. (we are setting a limit without clarifying it on the UI)

But let's image we go with your solution: the slider we show on the group creation doesn't even make sense because the only allowed answers is 1. (the group would have only 3 members when they start doing proposals).

And in a group with 10 members what's wrong with setting `9` as threshold?

> . If I did, it would mean I'm a bad project manager. :-\

I think being a bad project manager would be to not to use the expertise in your team to the fullest, and trust their judgment.

**Comment by sandrina-p** [comment]
Why can't the threshold be `threshold = MIN(threshold, groupSize)`? And if it's bigger than `groupSize` we disallow (with a warning) the creation of new proposals until the `threshold` is adjusted.

**Comment by taoeffect** [comment]
> We told you other solutions and each one has cons and pros.

It's not a solution if it breaks the proposals system, it's a bug.

> So, your solution can't go neither because doing MIN(threshold, half-of-group) is confusing as @mmbotelho mentioned before. (we are setting a limit without clarifying it on the UI)

So then let's clarify it in the UI?

> But let's image we go with your solution: the slider we show on the group creation doesn't even make sense because the only allowed answers is 1. (the group would have only 3 members when they start doing proposals).

Let's clear this confusion up: for the maximum value on the slider, it would be half of the `GROUP_LIMIT`, not half of the group size.

> And in a group with 10 members what's wrong with setting 9 as threshold?

> Why can't the threshold be `threshold = MIN(threshold, groupSize)`?

We could do that, maybe with the adjustment that it be `threshold = MIN(threshold, groupSize - 1)`. I'm OK with this solution because at least it means there's some possibility that the proposals system can still work.

> And if it's bigger than groupSize we disallow (with a warning) the creation of new proposals until the threshold is adjusted.

~~We can't do that, because it would mean on group creation they wouldn't be able to set it past 1.~~

EDIT: sorry, misread/misunderstood your comment. The issue with preventing proposals unless the threshold is adjusted, is that [you need a proposal to adjust the threshold](#issuecomment-613184751).

**Comment by taoeffect** [comment]
> I think being a bad project manager would be to not to use the expertise in your team to the fullest, and trust their judgment.

Heh, I agree. But I must trust my own judgement too, I cannot discount it, just as I cannot discount yours. :)

**Comment by taoeffect** [comment]
Just to be clear about this:

> We could do that, maybe with the adjustment that it be `threshold = MIN(threshold, groupSize - 1)`. I'm OK with this solution because at least it means there's some possibility that the proposals system can still work.

If we do it that way, it would mean that we leave open the possibility that a proposal still passes if everyone votes no, but a single person votes yes. I think that could result in a bad UX, but as @mmbotelho said, this is equivalent to setting an extremely low threshold, and letting users stab themselves.

If we do it this way, the two systems will be equivalent in the amount of damage they allow users to inflict upon themselves (but they remain very different types of voting systems, conceptually and mathematically).

On the plus side, that would mean they would allow equal amount of self-sabotage.

On the down side, that would mean they would allow equal amount of self-sabotage, and we wouldn't get to really test what happens in groups where they use a system that allows self-sabotage, and one that doesn't.

**Comment by sandrina-p** [comment]
Honestly... if I see myself in a group that I don't trust (because there's some kind of self-sabotage) I would just leave the group and start over with a different group that I trust.

I think we are doing too much negative assumptions about the people behind these (familiar/friendly) groups.

**Comment by taoeffect** [comment]
OK, well I'm OK with either:

- `threshold = MIN(threshold, currentGroupSize / 2)`
- `threshold = MIN(threshold, currentGroupSize - 1)`

Does that sound reasonable? Does anyone have a preference? I can post a poll to Slack if you'd like to vote on it. Or we can try to agree here. Either way, the UI will need to clarify that this is how it works. Figuring out the wording will be an interesting challenge.

**Comment by dotmacro** [comment]
Just as a reminder: the voting system and threshold are both set during group creation, when the group size is 1. As @sandrina-p reminded us above, proposals exist in groups with as few as 3 members, up to (of course) `GROUP_LIMIT`.

I'd like to suggest `threshold = MIN(threshold, currentGroupSize - 2)` (i.e. two yes votes) because in Robert's Rules every motion requires a second*, and two yes votes shows some agreement. My 2nd choice is `threshold = MIN(threshold, currentGroupSize - 1)`.

*Source: afaik, the only thing in Robert's Rules that doesn't require [a second](https://en.wikipedia.org/wiki/Second_%28parliamentary_procedure%29) is [an objection](https://en.wikipedia.org/wiki/Objection_to_the_consideration_of_a_question).

**Comment by taoeffect** [comment]
> I'd like to suggest `threshold = MIN(threshold, currentGroupSize - 2)` (i.e. two yes votes) because in Robert's Rules every motion requires a second*, and two yes votes shows some agreement. My 2nd choice is `threshold = MIN(threshold, currentGroupSize - 1)`.

OK, on exploring this, I'm realizing the math can get a bit confusing to me, so let's create a few user stories and see what happens when we use different voting rules, and different adjustments for `RULE_DISAGREEMENT`.

First, let's start by looking at what happens in `RULE_PERCENTAGE`.

### Example 1: Percentage(80%) / GroupSize(10) / 8Y / 2N

- Threshold: 80% of group size
- Group size: 10
- Yes votes: 8
- No votes: 2
- Outcome: **pass** (correct)

We expect the proposal to pass in this case, since the "threshold has been met", and indeed that is what the code shows:

```js
    const neededToPass = Math.ceil(threshold * (population - totalIndifferent))
    if (totalFor >= neededToPass) {
      return VOTE_FOR
    }
```

And in `RULE_DISAGREEMENT`, a proposal is killed as follows:

```js
    if (totalAgainst >= threshold) {
      return VOTE_AGAINST
    }
```

### Example 2: Disagreement(2) / GroupSize(10) / 8Y / 2N

- Threshold: 2 "no"s will block a proposal
- Group size: 10
- Yes votes: 8
- No votes: 2
- Outcome: **fail** (*correct* with our current broken rule)

`RULE_DISAGREEMENT` is biased in favor of blocking proposals when the "threshold is met".

How does this affect `threshold = MIN(threshold, currentGroupSize - 2)` ?

In this case `threshold = MIN(2, 8) = 2`, so we could have 8 "yes"s, and 2 "no"s, and the proposal will still be killed, as we expect. But, what happens in a different situation?

### Example 3: Disagreement(2) / GroupSize(3) / 2Y / 1N

- Threshold: 2 "no"s will block a proposal
- Group size: 3
- Yes votes: 2
- No votes: 1
- Outcome: **fail** (*incorrect* if `threshold = MIN(2, 3 - 2) = 1`)

Here we expected our proposal to pass, because it received only 1 objection, and we said we needed 2 objections to kill it.

This proposal should have passed, because only 1 "No" vote was received, and 2 are needed to kill the proposal. So this doesn't work.

What happens if we use `threshold = MIN(threshold, currentGroupSize - 1)`?

### Example 4: Disagreement(2) / GroupSize(3) / 2Y / 1N

- Threshold: 2 "no"s will block a proposal
- Group size: 3
- Yes votes: 2
- No votes: 1
- Outcome: **pass** (*correct* if `threshold = MIN(2, 3 - 1) = 2`)

Here the proposal passes, as expected. Would we encounter a similar problem if the group size was 2 using the same rule?

### Example 5: Disagreement(2) / GroupSize(2) / 1Y / 1N

- Threshold: 2 "no"s will block a proposal
- Group size: 2
- Yes votes: 1
- No votes: 1
- Outcome: **fail** (*incorrect (maybe?)* if `threshold = MIN(2, 2 - 1) = 1`)

EDIT: note that proposals for inviting/removing members don't happen at a group size of 2, but other proposal types might still exist, so leaving this example up.

The proposal failed, but according to the threshold setting of 2, this is wrong and @sandrina-p's rule of `threshold = MIN(threshold, currentGroupSize)` looks like it may be more intuitive in this case.

In other words, if you and the only other person in the group disagree as to whether a proposal should pass, the proposal will *pass* under a threshold of `2` using Sandrina's rule, and it will *fail* if using `2` for the threshold and the `MIN(threshold, groupSize - 1)` adjustment.

In @sandrina-p's rule, this is expected, and unexpected if you think that proposals should require at least two people agreeing to pass ("Robert's Rules" principle).

EDIT: Let's do one last example:

### Example 6: Disagreement(3) / GroupSize(3) / 1Y / 2N

- Threshold: 3 "no"s will block a proposal
- Group size: 3
- Yes votes: 1
- No votes: 2
- Outcome: **pass** (*???* if `threshold = MIN(3, 3) = 3`)

Using @sandrina-p's rule, the proposal passes even though 67% of the group disagreed, because the disagreement threshold was set to 3. @sandrina-p just double-checking, are you OK with this?

Under the "minus one" rule, the proposal would have failed, and under the divide-by-two rule, it definitely would have failed (but that rule's behavior might appear less intuitive with larger group sizes).

I think I am OK with this. Meaning, I am OK with either of these rules:

- `threshold = MIN(threshold, currentGroupSize - 1)`
- `threshold = MIN(threshold, currentGroupSize)`

Maybe even slightly in favor of `threshold = MIN(threshold, currentGroupSize)` as @sandrina-p suggested, even though in Example 6 it's slightly strange to me that a proposal passes even though most people disagreed with it. I'm OK with doing `threshold = MIN(threshold, currentGroupSize)` though.

@mmbotelho @dotmacro are you OK with using `threshold = MIN(threshold, currentGroupSize)`?

(Are there any weird cases I'm missing?)

**Comment by dotmacro** [comment]
> EDIT: note that proposals for inviting/removing members don't happen at a group size of 2, but other proposal types might still exist, so leaving this example up.

My understanding was we'd decided no proposals in groups of 1 because they'd always be unanimous, and no proposals in groups of 2 because they'd either be unanimous or complete disagreement. I can't find the issue now (maybe it's a Slack convo from before November?), so not certain of the details.

> Example 3

D'oh, right. 
Yes, `threshold = MIN(threshold, currentGroupSize - 1)` is clearly what I intended. Thanks for working through the maths, @taoeffect.

> (Are there any weird cases I'm missing?)

Yes, those with larger groups. 

### Example 7: Disagreement(30) / GroupSize(10) / 1Y / 9N

- Threshold: 30 "no"s will block a proposal
- Group size: 10
- Yes votes: 1
- No votes: 9
- Outcome: pass (if `threshold = MIN(30, 10) = 10`)

### Example 8: Disagreement(30) / GroupSize(10) / 1Y / 9N

- Threshold: 30 "no"s will block a proposal
- Group size: 10
- Yes votes: 1
- No votes: 9
- Outcome: fail (if `threshold = MIN(30, 10 - 1) = 9`)

or lower %-thresholds

### Example 9: Percentage(20%) / GroupSize(5) / 1Y / 4N

- Threshold: 20% "yes"s will pass a proposal
- Group size: 5
- Yes votes: 1
- No votes: 4
- Outcome: pass (if `threshold = .2 * 5 = 1`)

### Example 10: Percentage(20%) / GroupSize(5) / 1Y / 4N

- Threshold: 20% "yes"s will pass a proposal
- Group size: 5
- Yes votes: 1
- No votes: 4
- Outcome: fail (if `threshold = MAX(.2 * 5, 2) = 2`)

**Comment by taoeffect** [comment]
```
Yes, those with larger groups. 

### Example 7: Disagreement(30) / GroupSize(10) / 1Y / 9N

- Threshold: 30 "no"s will block a proposal
- Group size: 10
- Yes votes: 1
- No votes: 9
- Outcome: pass (if `threshold = MIN(30, 10)` = 10)
```

Right, it would require everyone to vote no in order to block it, because the threshold is set so high.

I'm guessing @sandrina-p would argue that this makes sense, because they set the threshold to 30, and so they meant for this to happen.

I can see that argument. I can also see how this would be confusing and the divide-groupSize-by-two rule might make it seem more intuitive to some users. I do see the difficulty in explaining in the UI how such a rule works. So I'm still leaning toward @sandrina-p's rule... slightly, followed by `MIN(threshold, currentGroupSize - 1)`, knowing that there might be confusion no matter which route we go.

**Comment by taoeffect** [comment]
A related problem with `RULE_DISAGREEMENT` is, unlike `RULE_PERCENTAGE`, we don't know when to throw up the warning that `RULE_PERCENTAGE` has for less-than-supermajority, when creating the group for the first time, because we don't know how many members the group is going to have.

Do we do it at 3? At 4? At 5?

**Comment by dotmacro** [comment]
I still strongly prefer `threshold = MIN(threshold, currentGroupSize - 1)`.

Requiring a minimum of 2 yes votes acts as a form of security, like 2FA. E.g. it prevents a hacker from taking over the group with merely one account because someone else would need to agree with anything they proposed.

We can apply the same 2-yes-minimum rule to both voting systems, which may even make the explanation a little easier.

**Comment by taoeffect** [comment]
> We can apply the same 2-yes-minimum rule to both voting systems, which may even make the explanation a little easier.

Well, here's another scenario: say everyone in your group has abandoned it, but not removed themselves from it. In this case you're the last active member, and you want to change the settings. You are the only "yes" vote that exists. At time of expiry, I think your proposal should pass if there were no other votes. So this scenario makes me more in favor of allowing proposals to pass with only 1 "yes" vote.

**Comment by mmbotelho** [comment]
I really like @sandrina-p's suggestion of using `threshold=MIN(threshold, currentGroupSize - 1)` because it's the perfect compromise between what users intended and what doesn't break the app!

Like you said @taoeffect we are also consistent in how much we let users "stab themselves" if they really want to!

**Comment by taoeffect** [comment]
So, it seems we're settling on:

```
threshold = MIN(threshold, currentGroupSize - 1)
```

Any objections? 😄

**Comment by dotmacro** [comment]
> Well, here's another scenario: say everyone in your group has abandoned it, but not removed themselves from it. In this case you're the last active member, and you want to change the settings. You are the only "yes" vote that exists. At time of expiry, I think your proposal should pass if there were no other votes. So this scenario makes me more in favor of allowing proposals to pass with only 1 "yes" vote.

In that case, the single remaining active user has a few options:
1. leave the group
2. continue as-is
3. contact other members and 
a) convince at least 1 to vote yes on an important proposal or 
b) realize that they should instead choose option 1 or 2

I'm unable to imagine a scenario in which one of those options won't work. Let me know if you think of one!

I prefer `threshold=MIN(threshold, currentGroupSize - 1)` for reasons stated, but I'm also ok with `threshold = MIN(threshold, currentGroupSize)` if that's what everyone decides :)

**Comment by taoeffect** [comment]
Commented at the same time. :)

Let's leave this up for 24 hours. If no objections, we'll go with `threshold = MIN(threshold, currentGroupSize - 1)`, and @mmbotelho I think that means you'll be able to close this issue?

**Comment by mmbotelho** [comment]
If there are no objections to using `threshold = MIN(threshold, currentGroupSize - 1)` I will take a look at Figma again and see if we need to do any changes to the mockups. I'll update you here!

**Comment by taoeffect** [comment]
OK opened #901 and closing this issue.

If anyone feels there's a strong reason to re-open please either do so or let us know (depending on what GitHub perms let you do).

**Comment by mmbotelho** [comment]
I'm reopening this one because, as I said in a previous comment, there some things we need to add to the UI so users know what's happening. I'll work on this somewhere in the near future.

**Comment by mmbotelho** [comment]
OK, I added an information message to the Group Settings page when the disagreement number is higher than the group size. This message is only displayed in that situation:

<img width="599" alt="Captura de ecrã 2020-04-28, às 16 34 47" src="https://user-images.githubusercontent.com/22004566/80507155-6b6dfe80-896e-11ea-95d8-66b7eaff3e5a.png">

@sandrina-p @taoeffect @dotmacro

**Comment by taoeffect** [comment]
@mmbotelho I think you're heading in the right direction!

Here's some feedback based on how this would be technically implemented: a user is welcome to set a disagreement threshold that is higher than their group size, e.g. say their group size is currently 2, and they want a disagreement threshold of 3. That is fine. They are allowed to set that, because when their group size increases up to 4 or more, it will use 3 as the threshold.

In the UI, we should also make sure to display the value as "3" if they set it to that, regardless of what it is.

However, we can do something like display next to it, an *effective threshold* (in grey? up to you), e.g. the setting is 3, but the effective threshold, because of the group size, might be 2, etc.

What do you think?

**Comment by taoeffect** [comment]
In other words, it's unclear to me from your screenshot, whether the blue text is saying that the number "1" was modified from whatever they set it to, or ... what.

**Comment by sandrina-p** [comment]
I'll be working on the implementation this week. There will be 2 PR: One with the layout only and another after to add the logic and tests.

**Comment by mmbotelho** [comment]
<img width="619" alt="Captura de ecrã 2020-04-29, às 12 48 14" src="https://user-images.githubusercontent.com/22004566/80592674-ba23a300-8a17-11ea-98c0-1dfeeb1b69de.png">

Better @taoeffect ?

**Comment by taoeffect** [comment]
@mmbotelho If we do it that way, users will think that they need to create a proposal to adjust the rule back to what they wanted it to originally be.

So I think it would be better to have the numbers reversed: have bold number be the actual number set/saved, and the text in grey showing the effective number.

**Comment by mmbotelho** [comment]
<img width="685" alt="Captura de ecrã 2020-04-29, às 18 00 45" src="https://user-images.githubusercontent.com/22004566/80624297-5e6f0f00-8a43-11ea-96cd-b8a6768d41df.png">

Like this? @taoeffect

**Comment by taoeffect** [comment]
@mmbotelho Yep! That looks great to me! 😄 👍

**Comment by dotmacro** [comment]
@mmbotelho @taoeffect Can the info text say "... your group is currently smaller..."?

**Comment by taoeffect** [comment]
Yeah adding "currently" makes sense to me 👍

**Comment by mmbotelho** [comment]
Changed the copy on Figma @dotmacro !

**Comment by sandrina-p** [comment]
Hi everyone. While implementing this, I'm feeling overwhelmed with all the naming we give to stuff related to this task. This comment will be just me trying to gather all the thoughts. Like a Glossary. You can ignore, or correct me if I got something wrong. 

- **Voting System**: The name given to this concept (on UI perspective).
- **`group.settings.proposals`** - Where all the data about this is stored in the code.
- **`group.settings.proposals[proposal_type][rule]`** - This `rule` is the voting system / proposal rule chosen by the user. It was decided to apply the same voting settings to all proposals types. There are two rules/systems:
  - **disagreement:** Define the number of people required to block a proposal.
  - **percentage:** Define the percentage of members who will need to agree to a proposal.
- **threshold**: The value defined for each rule/system. Disagreement: 1 to 60. Percentage: 0% to 100% (on the code `0.0` to `1.0`)
- **`PROPOSAL_PROPOSAL_SETTING_CHANGE:`** This is the name given to the proposal that changes the "Voting System". Hum... We refer to it as "voting system" and "proposal setting". It's the same thing, but with two different names. 
- **`contracts/voting/rules.js`** Where all the logic about "voting rules" / "voting system" is stored.

Regarding files, we have a little of everything: `GroupRules`, `GroupVotingSystem`, `ChangeVotingProposal`. Definitely, we should stick with one convention... Still wondering which one is the best.

Note: I'll update this as I progress in the task development...

Now, that I wrote this down, I see some room for improvement. Maybe we could rename some stuff so that the design and the code speak the same language. What do you think @taoeffect ?

**Comment by taoeffect** [comment]
> group.settings.proposals[rule]

No this would be `group.settings.proposals[proposal]`. The `rule` is in `settings.proposals[proposal].rule`, where `proposal` is one of the proposal constants, e.g. to take from your example, `PROPOSAL_PROPOSAL_SETTING_CHANGE`.

Example (taken from `Mincome.vue`):

```js
      const proposalRule = this.groupSettings.proposals[PROPOSAL_GROUP_SETTING_CHANGE]
      const { threshold } = proposalRule.ruleSettings[proposalRule.rule]
```

>  Maybe we could rename some stuff so that the design and the code speak the same language.

You are welcome to rename some of the `.vue` files as you see fit, but the core logic files (e.g. `proposals.js` and `rules.js`) shouldn't be renamed without significant deliberation and thought.

**Comment by sandrina-p** [comment]
`settings.proposals[proposal].rule` was a typo 🙈 (edited the original comment)

I'll do my best to rename Vue files to match the core logic in a better way.
